File: docs/validation/python/app/scenarios/s03_read_real.py
from pathlib import Path
from machine_config import MachineConfigReader

"""
S-03: Read real AconityMIDI fixture file
ID:          S-03
Title:       Read real-world AconityMIDI machine config file
Category:    happy-path
Layer:       reader
Precondition: Reference Materials/machine_config_TM_LPBF_02__AconityMIDI__OG_1783607045113 (1).h5
Action:      Parse the file, print all fields to stdout
Expected:    No error. All fields present in output match known machine parameters.
             (Record actual observed values in results.md — these become the
             reference for regression detection.)
Rationale:   This is the primary real-world validation. If the library cannot
             read a real file from the actual machine, it is not production-ready.
"""

def run(fixtures_dir: str, real_dir: str) -> tuple[bool, str]:
    matches = [
        p for p in Path(real_dir).glob("*.h5")
        if "AconityMIDI" in p.name and "OG_178" in p.name
    ]
    if not matches:
        return False, f"real AconityMIDI file not found in {real_dir}"

    cfg = MachineConfigReader(matches[0]).parse()
    fields = " | ".join([
        f"machine_name={cfg.meta.machine_name!r}",
        f"file_version={cfg.meta.file_version!r}",
        f"trains={len(cfg.optical_trains)}",
        f"build_plate_x={cfg.machine.build_plate.x}",
        f"build_plate_y={cfg.machine.build_plate.y}",
        f"wd={cfg.optical_trains[0].scanner.working_distance}",
        f"rotation[0]={cfg.optical_trains[0].scanner.scan_head_rotation}",
        f"hash={cfg.meta.configuration_hash[:16]}...",
    ])
    return True, fields

File: docs/validation/cpp/app/src/scenarios/s03_read_real.cpp
// S-03: Read real AconityMIDI fixture
#include "s03_read_real.hpp"
#include "machine_config/machine_config.hpp"

#include <sstream>

namespace s03 {

using namespace machine_config;

scenarios::Result run(const std::filesystem::path&, const std::filesystem::path& realDir) {
    std::filesystem::path matched;
    for (const auto& entry : std::filesystem::directory_iterator(realDir)) {
        auto name = entry.path().filename().string();
        if (entry.path().extension() == ".h5" &&
            name.find("AconityMIDI") != std::string::npos &&
            name.find("OG_178") != std::string::npos) {
            matched = entry.path();
            break;
        }
    }
    if (matched.empty()) {
        return {false, "real AconityMIDI file not found in " + realDir.string()};
    }

    try {
        MachineConfig cfg = MachineConfigReader{matched}.parse();
        std::ostringstream oss;
        oss << "machine_name=\"" << cfg.meta.machine_name << "\""
            << " | file_version=\"" << cfg.meta.file_version << "\""
            << " | trains=" << cfg.optical_trains.size()
            << " | build_plate_x=" << (cfg.machine.build_plate_x ? std::to_string(*cfg.machine.build_plate_x) : "null")
            << " | build_plate_y=" << (cfg.machine.build_plate_y ? std::to_string(*cfg.machine.build_plate_y) : "null")
            << " | wd=" << (cfg.optical_trains[0].scanner.working_distance ? std::to_string(*cfg.optical_trains[0].scanner.working_distance) : "null")
            << " | rotation[0]=" << (cfg.optical_trains[0].scanner.scan_head_rotation ? std::to_string(*cfg.optical_trains[0].scanner.scan_head_rotation) : "null")
            << " | hash=" << cfg.meta.configuration_hash.substr(0, 16) << "...";
        return {true, oss.str()};
    } catch (const std::exception& e) {
        return {false, std::string("read failed: ") + e.what()};
    }
}

} // namespace s03
